### PR TITLE
[gasoracle]feat: optimize how to query and calculate token ratio

### PR DESCRIPTION
**Pull request by Tri-stone**
# Goals of PR

Core changes:

- calculate eth_price, mnt_price from multi oracle, use `median` value of them
    - use `index price` from cex instead of latest price
- add bounds check for eth_price, mnt_price and token_radio
    - token_ratio = eth_price / mnt_price, and its bounds is derived from last token_ratio
        - token_ratio: [tokenRatioMin, tokenRatioMax]
        - tokenRatioMin := max(last_token_ratio*0.95, TokenRatioMin)
        - tokenRatioMax := min(last_token_ratio*1.05, TokenRatioMax)
        - TokenRatioMin: 100
        - TokenRatioMax: 10000
    - eth_price: (100, 1000000)
    - mnt_price: (0.01, 100)
- add `TOKEN_PAIR_MNT_MODE` to choose use `BIT` or `MNT` to calculate `token_ratio`

Related Issues:

- https://github.com/mantlenetworkio/mantle/issues/1054

Related PRs:
- https://github.com/mantlenetworkio/mantle/pull/1062

### CODE REVIEW

File: gas-oracle/tokenprice/tokenprice.go [hunk 1]
```diff
@@ -135,68 +154,74 @@ func (c *Client) PriceRatioWithMode() (float64, error) {
...
+	// calculate [tokenRatioMin, tokenRatioMax]
+	tokenRatioMin := getMax(c.lastRatio*0.95, TokenRatioMin)
+	tokenRatioMax := getMin(c.lastRatio*1.05, TokenRatioMax)
+
+	ratio := ethPrice / mntPrice
+	if ratio <= tokenRatioMin {
+		return tokenRatioMin
+	}
+	if ratio >= tokenRatioMax {
+		return tokenRatioMax
```
Comment (wukongcheng) [inline gas-oracle/tokenprice/tokenprice.go:196]: Where is the default value TokenRatio (t) = 4000 (from Ming’s spec) set?

Comment (Tri-stone) [reply gas-oracle/tokenprice/tokenprice.go:197]: In the code `gas-oracle/oracle/l1_client.go`, if `PriceRatioWithMode` returns an error, use the default value.

File: gas-oracle/tokenprice/tokenprice.go [hunk 2]
```diff
@@ -207,7 +208,7 @@ func determineTokenPairForMNT(tokenPairMNTMode bool) string {
 	}
 }
 
-func getMedian(nums ...float64) float64 {
+func getMedian(nums []float64) float64 {
 	sort.Float64s(nums)
 	return nums[len(nums)/2]
 }
```
Comment (wukongcheng) [inline gas-oracle/tokenprice/tokenprice.go:219]: Would directly iterating through the list to find the median be more efficient than sorting the list first and then retrieving the element at the middle position?

**Review by wukongcheng — APPROVED** [review]

**Review by bradyjoestar — APPROVED** [review]

**Comment by bradyjoestar** [comment]
LGTM.

**Review by Sha3nS — APPROVED** [review]

**Comment by Sha3nS** [comment]
LGTM

**Pull request merged**